### DISCUSSION TITLE
How to choose imgsz for best performance during training and validation?

### DISCUSSION THREAD

**Discussion by stealthgrip** [ultralytics/ultralytics#20400] [Q&A] [answered]
I am already some how familiar with basics of training and configuring YOLO models but there is a question I would like to hear professional opinion about.

When training a custom dataset, I see that we can set the **imgsz** parameter. I noticed training is faster at lower **imgsz**, but results seem worse. Could somebody please explain how to select the optimal **imgsz**, and whether it’s better to keep it the same for both training and validation? Are there any best practices?

Maybe this question already appeared somewhere in discussions but I have found nothing regarding this topic, using search bar.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @stealthgrip, thank you for your thoughtful question and for being a part of the Ultralytics community! 🚀 You can find detailed guidance in our [Docs](https://docs.ultralytics.com/), including [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples. Many frequently asked questions and best practices are covered there, so it's a great place to start.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If you’re seeking advice on optimal training parameters like **imgsz**, please share any additional training logs or dataset specifics—this helps us provide more targeted insights. Also, check out our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) for valuable recommendations.

You’re welcome to join discussions and ask questions in real-time on [Discord](https://discord.com/invite/ultralytics) 🎧, explore [Discourse](https://community.ultralytics.com/) for in-depth threads, or connect with peers on our [Subreddit](https://reddit.com/r/Ultralytics).

## Upgrade

Make sure you're using the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO runs smoothly in these verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing, verifying YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) across macOS, Windows, and Ubuntu on a 24-hour and per-commit basis.

This is an automated response 🤖—an Ultralytics engineer will review your question and provide a more detailed answer soon!

**Comment by pow3rpi** [comment, marked as answer]
Hi @stealthgrip! Thank you for your question!

The imgsz defines the resolution the images are resized to before being fed into the network. You are absolutely right that lower imgsz speeds up training but may degrade accuracy, especially for small objects and vice versa.

Best practices I would list the following:

- Keep imgsz the same for training and validation to ensure consistent evaluation. As far as I know, in ultralytics it is enough to specify imgsz in training config, so it will be applied on your validation set as well.
- Common balanced default value is 640. For small objects you may also try/prefer 768/960/1280.
- Choose imgsz divisible by 32 due to YOLO ultralytics architecture.

In general, imgsz=640 worked in most cases for me, so I would recommend you to always start with it. Then, if you have poor performance quality, other methods don't help you to boost it and your dataset contains a lot of small/tiny objects, you can try higher imgsz values like 768/960/1280.

I hope my comment will help you in your experience!

**Reply by stealthgrip** [reply]
Thank you @pow3rpi!!! Now it is all clear!
